### PR TITLE
Adds Clarification on Why Specifically What You Said Is Morally Reprehensible

### PR DESCRIPTION
**Pull request by bobbah**
## About The Pull Request
Added an additional portion to the filter message to help clarify why a message was filtered from IC chat.
epic screenshot

I additionally added a system to ensure users are discouraged from abusing this system, as requested by Oranges, through kicking if hitting the censor over 25 times in 60 seconds.
image

Finally, I removed this filtering from mobs without clients, as Oranges agreed this was unnecessary.
image

## Why It's Good For The Game
Please oh god why are my messages being blocked???? where are the rules????? adminhelp?????

## Changelog
:cl:
add: When the server denies your free speech you are now given the privilege of knowing why.
add: When users abuse these new privileges excessively within 60 seconds they will be disconnected from the server.
tweak: IC chat censors now only apply to mobs with clients.
/:cl:

### CODE REVIEW

**Comment by Repukan** [comment]
Looks disruptive and ugly as hell

**Comment by Repukan** [comment]
Just like OP

**Comment by bobbahbrown** [comment]
interesting point -- the original response for any blocked ic message was literally
image

given the circumstances i believe this is an improvement

**Comment by MrPerson** [comment]
It's supposed to be disruptive.

**Comment by daklj** [comment]
one more step towards introducing censorship on tg
it's against the spirit of the server and the game
all this because 3 anime admins decided to go full nazi on a thing that has never been an issue before

**Comment by mrhugo13** [comment]
Just gib anyone who  D A R E S  use netspeak like the fucking  P E A S A N T S  they are

**Comment by habader** [comment]
how are admins going to ban people for netspeak in ic now wtf

**Comment by necromanceranne** [comment]
But my ban quota QwQ

**Comment by bobbahbrown** [comment]
also it's not that i support this policy but instead i just want the quality of life of being told why my message is being censored instead of having no feedback.

if the censor list gets convoluted then without this feature it could be an annoying process to figure out why i can't say something

**Comment by Yttriums** [comment]
Block racial slurs and have this display a history of the slur's usage.

**Comment by mrhugo13** [comment]
> 
> 
> Block racial slurs and have this display a history of the slur's usage.

The DB handling that would break within a day.

**Comment by Yttriums** [comment]
> > Block racial slurs and have this display a history of the slur's usage.
> 
> The DB handling that would break within a day.

Lighten the load by allowing Italian terms through

**Comment by ExcessiveUseOfCobblestone** [comment]
does findtext return a t/f or does it return what you're finding?

**Comment by bobbahbrown** [comment]
@ExcessiveUseOfCobblestone it returns the position of the term you are searching for but because of our use of regex it's not incredibly helpful

**Comment by ExcessiveUseOfCobblestone** [comment]
okay, I was just seeing if I could cut down on proc repeats.

Q2: Can I crash the server by having a super long message and forcing it to loop through every word like paper?

**Comment by bobbahbrown** [comment]
@ExcessiveUseOfCobblestone 

Unlikely as any (sanitized) message is truncated to 1024 characters, though I'm not 100% certain so maybe its a better question for @optimumtact or the likes

**Comment by ExcessiveUseOfCobblestone** [comment]
I'm thinking just spamming it with x x x x x x until you max it then delete enough for lol at the end. that's at least a 300+ loop to work with per time you say it (although it would be really obvious what you're doing)

**Comment by nervere** [comment]
>Block racial slurs and have this display a history of the slur's usage.

This is not an intended function of the filter and we will not be using it for this purpose.
The purpose is to block netspeak only and that is the way I will be configuring the system.

**Comment by ike709** [comment]
Y'know, looking back, to avoid using regex (and simplify the output this PR adds) you could probably just rewrite the whole IC filter check to do:
```
if(GLOB.in_character_filter.len && !forced)
	var/list/words = splittext(message, " ")
	for(var/word in words)
		if(word in GLOB.in_character_filter)
			to_chat(src, "<span class='warning'>Your message has been blocked! \"<b>[word] 
                        </b>\" is prohibited! Consider reviewing the server rules.</span>")
			return
```

Although I'm not sure if this would be more or less expensive than the existing code.

**Comment by bobbahbrown** [comment]
> Y'know, looking back, to avoid using regex (and simplify the output this PR adds) you could probably just rewrite the whole IC filter check to do:
> 
> ```
> if(GLOB.in_character_filter.len && !forced)
> 	var/list/words = splittext(message, " ")
> 	for(var/word in words)
> 		if(word in GLOB.in_character_filter)
> 			to_chat(src, "<span class='warning'>Your message has been blocked! \"<b>[word] 
>                         </b>\" is prohibited! Consider reviewing the server rules.</span>")
> 			return
> ```
> 
> Although I'm not sure if this would be more or less expensive than the existing code.

I already wrote a cheaper solution, just testing it now!

**Comment by bobbahbrown** [comment]
@ExcessiveUseOfCobblestone @ike709 

Recursive solution above should improve performance outside of cases where users explicitly try to abuse this with things like "lol lol lol lol lol lol lol lol lol lol lol lol lol lol lol lol", but shouldn't really lower performance that much below split and loop solution in that worse case.

Proof of functionality:
image

**Comment by optimumtact** [comment]
I am very hesitant about adding this because it could become a dos vector

**Comment by optimumtact** [comment]
You should add some kind of counter and kick people who hit it in a 60 second period

**Comment by bobbahbrown** [comment]
on it chief

**Comment by JJRcop** [comment]
Can't you just use replacetext with the capture group

**Comment by bobbahbrown** [comment]
> Can't you just use replacetext with the capture group

findtext doesnt give you a capture group it gives you an index where the match starts

**Comment by bobbahbrown** [comment]
@optimumtact Implemented safety
image

**Comment by bobbahbrown** [comment]
Also updated the PR's comment to reflect the addition of security features as well as the removal of filtering from mobs without clients.

**Comment by JJRcop** [comment]
Mobs without client's can't speak without being forced to?

File: code/modules/mob/living/say.dm [hunk 1]
```diff
@@ -93,9 +113,26 @@ GLOBAL_LIST_INIT(department_radio_keys, list(
...
+			// Catch potential baddies
+			var/client/C = src.client
+			C.censoredMessageCounter++
+			if (C.censoredMessageCounter > CENSOR_MESSAGE_LIMIT)
+				to_chat(C, "<span class='userdanger'>You have exceeded the limit of [CENSOR_MESSAGE_LIMIT] censored messages in 60 seconds and have been disconnected. Administrators have been notified of this event.</span><br><span class='danger'>You may reconnect via the button in the file menu or by <b><u><a href='byond://winset?command=.reconnect'>clicking here to reconnect</a></u></b>.</span>")
+				QDEL_IN(C, 1) //to ensure they get our message before getting disconnected
+				message_admins("[key_name(src)] has been kicked for exceeding the limit of [CENSOR_MESSAGE_LIMIT] censored messages in 60 seconds.")
+			else if (C.censoredMessageCounter > CENSOR_MESSAGE_LIMIT - 5)
+				to_chat(C, "<span class='warning'>You are approaching the limit of [CENSOR_MESSAGE_LIMIT] censored messages in 60 seconds. You will be disconnected if you exceed this limit.</span>")
+			else if (C.censoredMessageCounter == 0)
```
Comment (optimumtact) [inline code/modules/mob/living/say.dm]: this will never be true because you always increment it first

Comment (bobbahbrown) [reply code/modules/mob/living/say.dm]: GOOD point i changed the position of that line briefly before committing one sec

Comment (bobbahbrown) [reply code/modules/mob/living/say.dm]: resolved in https://github.com/tgstation/tgstation/pull/44712/commits/546b661e57065fa4874f5279a209a765da8682df

**Comment by bobbahbrown** [comment]
JJRcop made a much more reasonable way of doing this so I'm closing this PR in favor of his #44728

**Pull request closed without merging**